protobuf/plugin/deepcopy/deepcopy.go

```diff
@@ -22,201 +22,250 @@ func (d *deepCopyGen) Name() string {
 }
 
 func (d *deepCopyGen) Init(g *generator.Generator) {
-	d.gen = g
+	d.Generator = g
 }
 
-func (d *deepCopyGen) genMsgDeepCopy(m *generator.Descriptor) {
-	if !plugin.DeepcopyEnabled(m.Options) {
-		return
-	}
+func (d *deepCopyGen) genCopyFunc(dst, src string) {
+	d.P(d.copyPkg.Use(), ".Copy(", dst, ", ", src, ")")
+}
 
+func (d *deepCopyGen) genMsgDeepCopy(m *generator.Descriptor) {
 	ccTypeName := generator.CamelCaseSlice(m.TypeName())
-	d.gen.P("func (m *", ccTypeName, ") Copy() *", ccTypeName, "{")
-	d.gen.P("\tif m == nil {")
-	d.gen.P("\t\treturn nil")
-	d.gen.P("\t}")
-	d.gen.P()
 
-	d.gen.P("\to := &", ccTypeName, "{")
+	// Generate backwards compatible, type-safe Copy() function.
+	d.P("func (m *", ccTypeName, ") Copy() *", ccTypeName, "{")
+	d.In()
+	d.P("o := &", ccTypeName, "{}")
+	d.P("o.CopyFrom(m)")
+	d.P("return o")
+	d.Out()
+	d.P("}")
+	d.P()
+
+	if len(m.Field) == 0 {
+		d.P("func (m *", ccTypeName, ") CopyFrom(src interface{})", " {}")
+		return
+	}
 
-	var funcs []func()
-	oneOfFuncs := make(map[string][]func())
+	d.P("func (m *", ccTypeName, ") CopyFrom(src interface{})", " {")
+	d.In()
+	d.P("if m == nil && src == nil {")
```

Comment:
Calling `CopyFrom` on a nil object should probably panic. If `src` is nil, we should probably clear the object.